internal/pkg/git.go

```diff
@@ -0,0 +1,220 @@
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"runtime"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/logging"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/version"
+)
+
+// GitSource deals with downloading individual files from a specific git repository over HTTPS
+type GitSource interface {
+	// Download fetches an io.ReadCloser for the schema file to download and also returns the size of the response (if known).
+	Download(
+		commit string,
+		getHTTPResponse func(*http.Request) (io.ReadCloser, int64, error)) (io.ReadCloser, int64, error)
+}
+
+// githubSource can download a plugin from github releases
+type githubSource struct {
+	host         string
+	organization string
+	repository   string
+	name         string
+
+	token string
+}
+
+// Creates a new github source adding authentication data in the environment, if it exists
+func newGithubSource(url *url.URL, name string) (*githubSource, error) {
+	contract.Requiref(url.Scheme == "github", "url", `scheme must be "github", was %q`, url.Scheme)
+
+	host := url.Host
+	parts := strings.Split(strings.Trim(url.Path, "/"), "/")
+
+	if host == "" {
+		return nil, fmt.Errorf("github:// url must have a host part, was: %s", url)
+	}
+
+	if len(parts) != 1 && len(parts) != 2 {
+		return nil, fmt.Errorf(
+			"github:// url must have the format <host>/<organization>[/<repository>], was: %s",
+			url)
+	}
+
+	organization := parts[0]
+	if organization == "" {
+		return nil, fmt.Errorf(
+			"github:// url must have the format <host>/<organization>[/<repository>], was: %s",
+			url)
+	}
+
+	repository := "pulumi-" + name
+	if len(parts) == 2 {
+		repository = parts[1]
+	}
+
+	return &githubSource{
+		host:         host,
+		organization: organization,
+		repository:   repository,
+		name:         name,
+
+		token: os.Getenv("GITHUB_TOKEN"),
+	}, nil
+}
+
+func (source *githubSource) newHTTPRequest(url, accept string) (*http.Request, error) {
+	var authorization string
+	if source.token != "" {
+		authorization = fmt.Sprintf("token %s", source.token)
+	}
+
+	req, err := buildHTTPRequest(url, authorization)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Accept", accept)
+	return req, nil
+}
+
+func (source *githubSource) getHTTPResponse(
+	getHTTPResponse func(*http.Request) (io.ReadCloser, int64, error),
+	req *http.Request,
+) (io.ReadCloser, int64, error) {
+	resp, length, err := getHTTPResponse(req)
+	if err == nil {
+		return resp, length, nil
+	}
+
+	// Wrap 403 rate limit errors with a more helpful message.
+	var downErr *downloadError
+	if !errors.As(err, &downErr) || downErr.code != 403 {
+		return nil, -1, err
+	}
+
+	// This is a rate limiting error only if x-ratelimit-remaining is 0.
+	// https://docs.github.com/en/rest/overview/resources-in-the-rest-api?apiVersion=2022-11-28#exceeding-the-rate-limit
+	if downErr.header.Get("x-ratelimit-remaining") != "0" {
```

Comment:
@AaronFriel this is code taken from core Pulumi (plugin releases download). Wouldn't it be overkill to include the Github Go SDK, given this tool only downloads the schema file during provider builds? During such builds we have access to a `GITHUB_TOKEN` for authorized downloads with much higher rate limits. I doubt we would bump into rate limits during provider builds.

It makes more sense to implement it using the Github Go SDK in the Pulumi CLI, so that all users have a better codebase for plugin releases downloads.

Reply:
That makes sense to me.